### PR TITLE
Fixes #3879 #4283  Other pictures flickers when a media is selected from contributions list

### LINKED ISSUE #3879: Other pictures flickers when a media is selected from contributions list
**Summary:** 
I always knew I was seeing flickering images when I choose an item from contrib list and trying to go to Media Details Fragment. While I was testing another bug, I recognized on click to an object from contrib list, rebuildCatList method from MediaDetailsFragment is called for previous images too. And I see previous image is appeared during my debug steps. 

This issue should be investigated.

### PR DESCRIPTION
**Pull request by vinayak0505**
**Description (required)**
I am sorry for previous pull request as i panicked and push the wrong one

Fixes #3879 #4283

What changes did you make and why?
I just added a feature to check it the clicked item is loaded before setting that item, which will remove image flicker as program will not wait for 5 milliseconds and random crashes

**Tests performed (required)**
manual testing

please check this i am really sorry about previous pull request @nicolas-raoul @madhurgupta10 @neslihanturan

### CODE REVIEW

**Comment by nicolas-raoul** [comment]
No problem!
By the way, you can edit a pull request simply by pushing to the same remote branch. No need to close the PR and open a new one :-)

**Comment by vinayak0505** [comment]
Thank you , I panicked a lot so I thought of starting from scratch , and
also know if you like we can add animation, maybe similar to the one when
you press notification bell

File: app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java [hunk 1]
```diff
@@ -415,4 +428,6 @@ public int getCount() {
             return provider.getTotalMediaCount();
         }
     }
+
+
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:432]: Would you mind undoing the addition of these two empty lines? Or was it on purpose maybe?

File: app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java [hunk 2]
```diff
@@ -332,13 +332,26 @@ private void updateBookmarkState(MenuItem item) {
...
-        handler.postDelayed(() -> pager.setCurrentItem(i), 5);
+        settingViewPagerCurrentItem(i);
     }
 
     public void showImage(int i) {
-        Handler handler =  new Handler();
-        handler.postDelayed(() -> pager.setCurrentItem(i), 5);
+        settingViewPagerCurrentItem(i);
+    }
+    private void settingViewPagerCurrentItem(int i) {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:341]: Since a few years our rule is that we need javadoc for every method addition. Even if self-explanatory. And yes, I know that most methods do not have javadoc yet, we are working on it :-)
Also, one empty line to separate it from the previous method.
Thanks!

File: app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java [hunk 3]
```diff
@@ -332,13 +332,26 @@ private void updateBookmarkState(MenuItem item) {
...
+        final Boolean[] currentItemNotShown = {true};
+        Runnable runnable = new Runnable() {
+            @Override
+            public void run() {
+                while(currentItemNotShown[0]){
+                    if(adapter.getCount()>i){
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:347]: space before and after `<`

File: app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java [hunk 4]
```diff
@@ -332,13 +332,26 @@ private void updateBookmarkState(MenuItem item) {
...
+                        pager.setCurrentItem(i,false);
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:348]: Space after `,`

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
It works great!
Tapping a contribution now leads to a very smooth experience (I mean there is no flickering).

I just asked for a few minor changes.
Thanks!

File: app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java (hunk 1 above)
Comment (vinayak0505) [reply app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:432]: no it was a mistake

**Comment by vinayak0505** [comment]
done

**Comment by neslihanturan** [comment]
When I was logged out and tried to display a media detail, it crashed with:
```
free.nrw.commons, PID: 29828
    java.lang.NullPointerException: Attempt to invoke virtual method 'int fr.free.nrw.commons.media.MediaDetailPagerFragment$MediaDetailAdapter.getCount()' on a null object reference
        at fr.free.nrw.commons.media.MediaDetailPagerFragment$1.run(MediaDetailPagerFragment.java:354)
        at java.lang.Thread.run(Thread.java:919)
2021-04-08 19:33:20.862 29828-4650/fr.free.nrw.commons E/ACRA: ACRA caught a NullPointerException for fr.free.nrw.commons
    java.lang.NullPointerException: Attempt to invoke virtual method 'int fr.free.nrw.commons.media.MediaDetailPagerFragment$MediaDetailAdapter.getCount()' on a null object reference
        at fr.free.nrw.commons.media.MediaDetailPagerFragment$1.run(MediaDetailPagerFragment.java:354)
        at java.lang.Thread.run(Thread.java:919)
```

**Comment by neslihanturan** [comment]
I works good on contributions list, however it crashes then it comes to explore fragment:
```
    java.lang.NullPointerException: Attempt to invoke virtual method 'int fr.free.nrw.commons.media.MediaDetailPagerFragment$MediaDetailAdapter.getCount()' on a null object reference
        at fr.free.nrw.commons.media.MediaDetailPagerFragment$1.run(MediaDetailPagerFragment.java:354)
        at java.lang.Thread.run(Thread.java:919)
2021-04-08 19:36:22.588 29830-5120/fr.free.nrw.commons E/ACRA: ACRA caught a NullPointerException for fr.free.nrw.commons
    java.lang.NullPointerException: Attempt to invoke virtual method 'int fr.free.nrw.commons.media.MediaDetailPagerFragment$MediaDetailAdapter.getCount()' on a null object reference
        at fr.free.nrw.commons.media.MediaDetailPagerFragment$1.run(MediaDetailPagerFragment.java:354)
        at java.lang.Thread.run(Thread.java:919)
        ```

File: app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java [hunk 5]
```diff
@@ -332,13 +332,33 @@ private void updateBookmarkState(MenuItem item) {
...
-        Handler handler =  new Handler();
-        handler.postDelayed(() -> pager.setCurrentItem(i), 5);
+        settingViewPagerCurrentItem(i);
+    }
+
+  /**
+   * This function is called to check if the selected item(item to be displayed) is loaded in the
+   * adapter or not, and when its done in background so that it does not block the ui if item is
+   * removed
+   * @param i current item that to be shown
```
Comment (neslihanturan) [inline app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:346]: I think this solution won't work because we don't use MediaDetailPagerFragment only for contribution list. We use it in explore, bookmarks and searchFragments too. So keeping an index is a too hacky and patchy way to solve this issue which would highly cause some problems for us. So I suggest an overall strategy change for this issue. To start we can discover the root cause of this issue.

**Review by neslihanturan — CHANGES_REQUESTED** [review]

**Comment by vinayak0505** [comment]
image
nice catch, as you can see here adapter is not initialized for first 100 millisecond which cases a problem so there are two ways to solve this if i wait for adapter to get initialize or the right way ill fix this adapter, ill try to go with 2nd option first but i have no idea why the programmer has done this and why this was allowed

**Comment by neslihanturan** [comment]
And welcome to the project @vinayak0505 , please do not get stressed if something goes wrong with the PR. It is always normal to close them to start from scratch. It is normal to send new commits to the existing PR too.

**Comment by neslihanturan** [comment]
> image
> nice catch, as you can see here adapter is not initialized for first 100 millisecond which cases a problem so there are two ways to solve this if i wait for adapter to get initialize or the right way ill fix this adapter, ill try to go with 2nd option first but i have no idea why the programmer has done this and why this was allowed



> image
> nice catch, as you can see here adapter is not initialized for first 100 millisecond which cases a problem so there are two ways to solve this if i wait for adapter to get initialize or the right way ill fix this adapter, ill try to go with 2nd option first but i have no idea why the programmer has done this and why this was allowed

I am not really sure why this delay is there, but considering this is a network operation, I assume it might be required. Anyway please share your further findings with us :)

**Comment by vinayak0505** [comment]
> And welcome to the project @vinayak0505 , please do not get stressed if something goes wrong with the PR. It is always normal to close them to start from scratch. It is normal to send new commits to the existing PR too. 

Thank you very much and about stressing that you just mentioned if it's because you read the above comments then thank you I felt really good but if something I might blame a mistake on others for my PR then don't worry If I have do a mistake I'll accept that even if I lose an opportunity

**Comment by vinayak0505** [comment]
hey!! thanks for the report but please create a different bug report for this as i have figured out what was wrong as problem was in the explore Fragment code i can solve this hear but then i wont get point for GSoC, if creating a issue would problematic then no issue ill upload it here only, ill make point from other bugs no issue @neslihanturan

File: app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java (hunk 5 above)
Comment (vinayak0505) [reply app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:346]: yes 
exactly 
i have done that just waiting for your reply to make commit

Comment (vinayak0505) [reply app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:346]: I don't know like who does that , someone has called the function made it wait for 5millisec and in mean time applied the adapter, like simply just apply the adapter one line above all problem solved , same can is happening in MediaDetailPagerFragment and 
image
luckily we can see that this used in 5 places probably with same error so those screen should be also flickering some times , that would be solved to

**Comment by neslihanturan** [comment]
> hey!! thanks for the report but please create a different bug report for this as i have figured out what was wrong as problem was in the explore Fragment code i can solve this hear but then i wont get point for GSoC, if creating a issue would problematic then no issue ill upload it here only, ill make point from other bugs no issue @neslihanturan

We don't only consider number of PRs, we consider their difficulty level for GSoC. There is not a second issue here, this issue covers all of the steps we talked about. So there is no need to create one more issue or one more PR. 

Can you please commit your new solution (it is hard to follow from code screen shoots, I prefer to see commits instead)?

Also, sometimes I am having hard times to understand your messages. Could you please make sure you follow punctuation rules? So that I could follow sentence ends and beginnings. Thanks!:)

**Comment by vinayak0505** [comment]
I am working on solving this. 

I'll take care of punctuation from next time.

**Comment by vinayak0505** [comment]
solved please review it.

**Review by nicolas-raoul — COMMENTED** [review]
Thanks for fixing the stylistic issues!

I just tested with this branch's latest code, it works great for me even when scrolling and tapping like crazy.

Neslihan, if you know a way to trigger the crash with higher probability (for instance by tapping fast in a particular order), please let us know :-)

**Comment by vinayak0505** [comment]
Can I start working on other issues? 
I have checked from my side as this function is used in five different places and I have checked them

**Comment by nicolas-raoul** [comment]
Yes, feel free to work on other issues while we test this branch further.
By the way, I suggest you create a new branch for each pull request, rather
than working directly on your master branch.

**Comment by vinayak0505** [comment]
@neslihanturan please review this :)

File: app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java [hunk 6]
```diff
@@ -332,13 +320,31 @@ private void updateBookmarkState(MenuItem item) {
...
-        Handler handler =  new Handler();
-        handler.postDelayed(() -> pager.setCurrentItem(i), 5);
+        settingViewPagerCurrentItem(i);
+    }
+
+  /**
+   * This function waits for the item to load then sets the item to current item
+   * @param position current item that to be shown
+   */
+    private void settingViewPagerCurrentItem(int position) {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:334]: I just noticed the method name. As you probably know, Java method names should begin with a verb, except for some rare exceptions. Do you have a special reason for not using a verb? If not, would you mind renaming it? Thanks!

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
I have been using this branch for 2 days, it is working great so far, I really appreciate that flickering (previous image shown briefly upon tapping) does not happen anymore.

Apart from the small detail I noticed right now, the code looks reasonable to me.

File: app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java (hunk 6 above)
Comment (vinayak0505) [reply app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java:334]: No, there is no specific reason for that. Even thought I always use verb, i wasn't aware about this hard and fast rule. Thanks for telling me.

**Comment by vinayak0505** [comment]
Done

**Review by nicolas-raoul — APPROVED** [review]

**Comment by vinayak0505** [comment]
@neslihanturan I have made changes as per your requirement. so please check.

**Comment by vinayak0505** [comment]
I have also made an upstream pull to my local branch to check for conflicts.

**Comment by vinayak0505** [comment]
I have merged upstream also please

**Comment by neslihanturan** [comment]
Thanks @vinayak0505 and @nicolas-raoul , I am on it :)

**Review by neslihanturan — CHANGES_REQUESTED** [review]
Hey @vinayak0505 there are several irrelevant changes made in this PR, please revert them. Only required lines to solve the issue should be changed. So when you switched to files tab on github (or when you check git diff) you should only see related lines are edited. Please make sure you create a new branches on each task, otherwise (if you work on master for all tasks) such problems may occur. You can follow our contributors guideline for information about git flows. Please ping me for review as soon as you solved this issue.

If you are confused feel free to close this PR and open a new one.

**Comment by vinayak0505** [comment]
@neslihanturan oblast 2 commits I only pull upstream to check I revert them back but they didn't updated can merge while leaving the last 2 commits

**Comment by vinayak0505** [comment]
Check the third last commit

**Comment by nicolas-raoul** [comment]
@vinayak0505 It is a good thing to keep PRs up-to-date, your intent was great.
To do so, better use Git's `rebase` operation, though. Would you mind trying?
If it does not work, an easy workaround is to create a new PR as Neslihan suggested.
Thanks!

**Comment by vinayak0505** [comment]
Oh okay, it would be okay if I try this for my next pr or should I change the pr now.

**Comment by nicolas-raoul** [comment]
I would be great if you could do it for this one :-)

**Comment by vinayak0505** [comment]
@nicolas-raoul I revert back the changes and made a rebase but then it shows more files changed so i created a new pull request

**Comment by vinayak0505** [comment]
@neslihanturan please check pull  request #4354
